### PR TITLE
[CWS] Compute workloads SBOMs to enrich events and activity dumps

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR introduces a way to generate SBOMs by calling a modified version of Trivy, directly from system-probe. This modified version of Trivy will fetch the files owned by each package so that we can resolve the package names and versions involved in a CWS event.

This new feature is experimental and controller with these new configuration parameters:
- runtime_security_config.sbom.enabled : controls if this feature should be enabled (set to false by default)
- runtime_security_config.sbom.workloads_cache_size : controls the size of the SBOM cache (indexed by image name and version) that we introduced to prevent the scanner from scanning the same container images over and over again.

### Motivation

This PR introduces powerful new features for CWS:
- the ability to write a rule that will trigger a CWS event as soon as a vulnerable package is loaded in memory
- new metadata on which to base a "risk score" for CWS's activity dumps feature

### Describe how to test/QA your changes

1) Add this rule to your policy:

```bash
 - id: package_test
   expression: exec.file.name != "" && exec.file.package.name == "bash"
```

2) Enable the configuration parameter and start a Debian container. Since the scanning is asynchronous (for performance & coverage reasons), wait a little bit and then run "bash". You should see an event trigger.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by kayayarai — APPROVED** [review]
Docs 👍

**Review by brycekahle — APPROVED** [review]
👍 for ebpf-platform owned files

File: pkg/security/resolvers/sbom/resolver.go [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+				sbom.files[file] = pkg
+			}
+		}
+	}
+
+	// we can get rid of the report now that we've generate the file mapping
+	sbom.report = nil
+
+	// mark the SBOM ass successful
+	sbom.scannSuccessful.Store(true)
```
Comment (YoannGh) [inline pkg/security/resolvers/sbom/resolver.go:278]: nit: typo

File: pkg/security/resolvers/cgroup/model.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	for _, pid := range pids {
+		newCGroup.PIDs.Add(pid, 0)
+	}
+	return &newCGroup, nil
+}
+
+// GetRootPIDs returns the list of root pids for the current workload
+func (cgce *Cgroup) GetRootPIDs() []uint32 {
```
Comment (Gui774ume) [inline pkg/security/resolvers/cgroup/model.go:48]: It's a nit but these aren't root pids anymore: these are all the pids inside this cgroup.

File: pkg/security/secl/model/model.go [hunk 3]
```diff
@@ -509,6 +509,12 @@ type FileEvent struct {
 
 	PathResolutionError error `field:"-" json:"-"`
 
+	PkgName    string `field:"package.name,handler:ResolvePackageName"`       // SECLDoc[package.name] Definition:`Name of the package that provided this file`
+	PkgVersion string `field:"package.version,handler:ResolvePackageVersion"` // SECLDoc[package.version] Definition:`Full version of the package that provided this file`
+	PkgMajor   int    `field:"package.major,handler:ResolvePackageMajor"`     // SECLDoc[package.major] Definition:`Major version of the package that provided this file`
```
Comment (YoannGh) [inline pkg/security/secl/model/model.go:514]: Remove those

File: pkg/security/resolvers/cgroup/model.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	cgce.Lock()
+	defer cgce.Unlock()
+
+	return cgce.PIDs.Keys()
+}
+
+// RemoveRootPID removes the provided root pid from the list of pids
+func (cgce *Cgroup) RemoveRootPID(pid uint32) {
```
Comment (Gui774ume) [inline pkg/security/resolvers/cgroup/model.go:56]: Same here

File: pkg/security/resolvers/cgroup/model.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/hashicorp/golang-lru/v2/simplelru"
+	"go.uber.org/atomic"
+)
+
+// Cgroup cgroup resolver cache entry
+type Cgroup struct {
+	sync.RWMutex
+	Deleted *atomic.Bool
+	ID      string
+	Tags    []string
```
Comment (Gui774ume) [inline pkg/security/resolvers/cgroup/model.go:24]: So do we need to store the tags in the end ?

File: pkg/security/probe/field_handlers.go [hunk 6]
```diff
@@ -442,3 +442,49 @@ func (fh *FieldHandlers) ResolveEventTimestamp(ev *model.Event) time.Time {
...
+			f.PkgName = pkg.Name
+			f.PkgVersion = pkg.Version
+		}
+	}
+	return f.PkgName
+}
+
+// ResolvePackageVersion resolves the version of the package providing this file
+func (fh *FieldHandlers) ResolvePackageVersion(ev *model.Event, f *model.FileEvent) string {
+	if f.PkgVersion == "" {
```
Comment (YoannGh) [inline pkg/security/probe/field_handlers.go:459]: Should we ensure that the path was fully resolved before resolving the package ?

File: pkg/security/resolvers/process/resolver.go [hunk 7]
```diff
@@ -544,7 +544,7 @@ func (p *Resolver) insertEntry(entry, prev *model.ProcessCacheEntry, origin proc
 		prev.Release()
 	}
 
-	if p.cgroupResolver != nil {
+	if p.cgroupResolver != nil && entry.ContainerID != "" {
```
Comment (Gui774ume) [inline pkg/security/resolvers/process/resolver.go:547]: Are you sure you want to filter out the host ? I know we have another filter later on to filter out the host, so maybe we should keep just one fail safe ?

**Review by paulcacheux — APPROVED** [review]

**Review by lebauce — APPROVED** [review]

**Review by YoannGh — APPROVED** [review]

**Review by amenasria — APPROVED** [review]
LGTM for Agent Platform owned files !

**Pull request merged**